File: sorting/heap_sort_recursive/heap_sort.cpp
#include <utility>

inline std::size_t parent_index(const std::size_t index) {
    return index / 2;
}

inline std::size_t left_index(const std::size_t index) {
    return index * 2;
}

inline std::size_t right_index(const std::size_t index) {
    return index * 2 + 1;
}

template <typename T>
void heapify(T *elements, const std::size_t parent, const std::size_t size) {
    std::size_t left = left_index(parent);
    std::size_t right = right_index(parent);
    std::size_t largest = parent;

    if (left < size && elements[parent] < elements[left]) {
        largest = left;
    }

    if (right < size && elements[largest] < elements[right]) {
        largest = right;
    }

    if (largest != parent) {
        std::swap(elements[parent], elements[largest]);
        heapify(elements, largest, size);
    }
}

template <typename T>
void heap_sort(T *elements, const std::size_t size) {
    if (elements == nullptr) {
        return;
    }

    auto heap_size = parent_index(size - 1);
    for (auto i = 0ul; i <= heap_size; i++) {
        heapify(elements, heap_size - i, size);
    }

    for (auto i = 0ul; i < size; i++) {
        auto index = size - i - 1;

        std::swap(elements[0], elements[index]);
        heapify(elements, 0, index);
    }
}

File: sorting/heap_sort_recursive/heap_sort.js
'use strict';

function swap(elements, index0, index1) {
	const tmp = elements[index0];
	elements[index0] = elements[index1];
	elements[index1] = tmp;
}

function parentIndex(index) {
	return Math.floor(index / 2);
}

function leftIndex(index) {
	return index * 2;
}

function rightIndex(index) {
	return index * 2 + 1;
}

function heapify(elements, parent, size) {
	const left = leftIndex(parent);
	const right = rightIndex(parent);
	let largest = parent;

	if (left < size && elements[parent] < elements[left]) {
		largest = left;
	}

	if (right < size && elements[largest] < elements[right]) {
		largest = right;
	}

	if (largest != parent) {
		swap(elements, parent, largest);
		heapify(elements, largest, size);
	}
}

function heapSort(elements) {
	const lastParent = parentIndex(elements.length - 1);
	for (let i = lastParent; i > 0; i--) {
		heapify(elements, i, elements.length);
	}

	for (let i = elements.length - 1; i > 0; i--) {
		swap(elements, 0, i);

		heapify(elements, 0, i);
	}
}

let a = [9, 8, 7, 6, 5, 4, 3, 2, 1, 0];
heapSort(a);

console.log(a);

